docs(dropdown): add UX description (#1258)
siemens/ix: packages/documentation/docs/controls/_dropdown_styleguide.md

Reviewer: I'm missing some of the properties:
* close behavior
* placement
* position strategy (?)
Haven't given it too much thought but I feel like these are "behavior in context" topics 😊

Dropdowns are containers into which you can insert elements. The dropdown containers do not function independently; they must be opened via a dropdown button, for example. The dropdown container contains items that offer the user a specific selection. These items can be customized depending on their intended use.

Caution, only the dropdown container and its items are described here! For information about the element that opens it, read the article about the dropdown button or the split button.

![Overview](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2353-2278&mode=design&t=OVHeXvLZYLkP2CzN-4)

1. Dropdown Container
2. Item
3. Header
4. Quick Actions
5. Submenu Item
6. Checked
7. Separator
8. Icon
9. Label

## Options

> - **Header:** Add an header for the dropdown container. 

Reviewer: keep an eye on the usage of an vs. a, and ending every item with a full stop (.)

- **Quick Actions:** Add a quick action bar.
- **Checked:** Mark selected items in the dropdown with a check mark.
- **Submenu:** Add an submenu for a multi-level dropdown.
- **Separator** Add a separator to visually divide items from each other 

> - **Icon:** Select, if required, a fitting icon for the item.

Reviewer: try to make this sentence less complicated (you could align with the wording from the button or pane chapter)

- **Label:** Set a label for the dropdown item. We typically use short labels including verbs.

Note: For a date selection dropdown, use the date dropdown!

![Dropdown Examples](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2372-2696&mode=design&t=OVHeXvLZYLkP2CzN-4)

## States

> Dropdown containers themselves have no defined stats - but their items have five: Default, hover, active, disabled and focused. In an active state, dropdown submenu items show an additional level with selectable options.

Reviewer: The second sentence needs to be rephrased, not sure if it's correct

![Item States](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2343-42235&mode=design&t=OVHeXvLZYLkP2CzN-4)

> ## Does and Don’ts

Reviewer: typo :)

- Do use dropdowns to showcase related actions

> - Don’t use too many dropdown items - we recommend 7 +/- 2

Reviewer: Maybe you can find some more dos and donts?

## Related patterns

> - [Dropdown button](dropdown-button.md)

Reviewer: since these patterns are in a folder, you need to add them to your path (I believe buttons/dropdown-button.md)

Dropdown containers display a menu with additional items as soon as a user clicks on the trigger element, e.g. a dropdown button. Clicking on one of the items in the dropdown performs the action. We typically use dropdowns to allow users to select one option from a list of available choices.

> For information about the trigger element that opens the dropdown, please refer to our [dropdown button](buttons/dropdown-button.md) and [split button](buttons/split-button.md) components.

Reviewer 2: Any element can be set as trigger for a dropdown. Dropdown button and split button are just two special cases that have the dropdown functionality baked into them.

Reviewer 2: ```suggestion
```

![Overview](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2353-2278&mode=design&t=OVHeXvLZYLkP2CzN-4)

1. Dropdown container
2. Item
3. Header
4. Quick actions
5. Submenu
6. Checked
7. Separator
8. Icon
9. Label

## Options

- **Header:** Add a header for the dropdown container. This allows users to better understand which elements are in the dropdown.
- **Quick Actions:** Add a quick action bar. Add 3 to 5 actions/items to the quick actions. We typically use quick actions to access common functions such as cut, copy and paste.
- **Checked:** Mark selected items in the dropdown with a check mark. We typically use check marks when an item can be activated (e.g. for the dropdown "Sort by:" selection: name, modified date, create date).
- **Submenu:** Add a submenu for a multi-level dropdown. We typically use a submenu when there is a long item list and different systematic categorization within the items.
- **Separator:** Add a separator to visually divide items from each other. We normally use a separator to isolate individual elements from a cohesive item list.

> - **Icon:** Icons can also be displayed with item labels. We typically use an icon in addition to the label when it has a recognizable benefit.

Reviewer 2: What is a 'recognizable benefit'?

Reviewer 3: Idea:
**Icon:** Icons can be displayed to support the label and make the item faster discoverable. The icon should be widely known for representing the action or function among your users .

Reviewer 2: ```suggestion
- **Icon:** Icons can be displayed to support the label and make the item more easy to discover by the user. The icon should be widely known for representing the action or function among your users.
```

- **Label:** Set a label for the dropdown item. We typically use short labels including verbs.
- **Trigger:** The trigger defines which element opens the dropdown. A trigger should also be defined for a dropdown submenu. We typically use a button as the trigger element.
- **Anchor:** An anchor defines where the dropdown is placed. When no anchor is defined the trigger element is used as the anchor.
- **Close Behavior:** Defines whether a click inside and/or outside the dropdown closes the dropdown. A submenu is always closed together with the parent dropdown. Three Options are possible: 
    - Inside: clicking within the dropdown closes the dropdown.
    - Outside: clicking outside the dropdown closes the dropdown.

>     - Both: clicking within and outside the dropdown closes the dropdown.

Reviewer 2: ```suggestion
    - Both: clicking within and outside the dropdown closes the dropdown.
    - false: dropdown will only close if it's parent gets closed
```

> - **Placement:** Place a dropdown at the top, bottom, left or right edge as well as at the beginning or end of the trigger element. The placement may be automatically adjusted in case it cannot be displayed correctly (detailed behavior described in the context). We typically use the default (bottom right) placement option to ensure consistency.

Reviewer 2: ```suggestion
- **Placement:** Place a dropdown at the top, bottom, left or right edge as well as at the beginning or end of the trigger/anchor element. The placement may be automatically adjusted in case it cannot be displayed correctly (detailed behavior described in the context section below). We typically use the default (bottom right) placement option to ensure consistency.
```

- **Date Selection:** Use the component [date dropdown](date-dropdown.md) to get a date selection in the dropdown.

![Dropdown Examples](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2372-2696&mode=design&t=OVHeXvLZYLkP2CzN-4)

## Behavior in context
- **Text truncation:** The labels of the items and the header only consist of one line. A truncation only occurs if there is not enough space on the screen.
- **Scrollbar:** A dropdown is provided with a scrollbar when the dropdown takes up 50% of the screen.

> - **Placement:** The position depends on the trigger element (e.g. button). By default, the dropdown is displayed at the bottom right of trigger element. When there is not enough space for the selected placement, it will be corrected automatically. The placement of the submenu is always generated automatically. 

Reviewer 2: ```suggestion
- **Placement:** The position depends on the trigger/anchor element (e.g. button). By default, the dropdown is displayed at the bottom right of trigger element. When there is not enough space for the selected placement, it will be corrected automatically. The placement of the submenu is always generated automatically. 
```

- **Quick actions:** Quick actions only consist of icons, therefore it is important to use icons that are understandable without a label or tooltip. A quick action bar can also be used without additional items in the dropdown. 

![Dropdown in Context](https://www.figma.com/design/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?node-id=2463-3302&t=QaiBJKNOwHMdBuk2-4)

## States
Dropdown items have five states: Default, hover, active, disabled and focused. When a submenu is in an active state, the submenu displays an additional dropdown with selectable options.

![Item States](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2343-42235&mode=design&t=OVHeXvLZYLkP2CzN-4)

## Dos and Don’ts
- Do structure the items in a dropdown in a coherent way; use submenus, quick actions and separators if beneficial
- Do use dropdowns to showcase related actions
- Do disable items that can't be used at the moment
- Don’t use global navigation options in a dropdown

> - Don’t use too many dropdown items - we recommend 7 +/- 2

Reviewer 2: ```suggestion
- Don’t use too many dropdown items - we recommend a maximum of seven
```

> Dropdown containers are elements that display a menu with additional items as soon as a user clicks on the top element, e.g. a dropdown button. Clicking on one of the items in the dropdown triggers the action. We typically use dropdowns to allow users to select one option from a list of available choices.

Reviewer: Maybe you could make this sentence more simple, like "Dropdowns display a menu ..."

> For information about the element that opens the dropdown containers, please refer to our separate [Dropdown button](dropdown-button.md) and [Split button](split-button.md) guide.

Reviewer: sounds better! I believe that it should be lowercase (dropdown and split button)
Plus, maybe "guide" is not the right word here - maybe "components"?

In general, I would stick to "dropdown" instead of "dropdown container" since this is actually the name of the component (we can discuss this in person 😉 )

Author: That makes sense. I followed the sentence in the style guide from the dropdown button, should we change the sentence too?

![Overview](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2353-2278&mode=design&t=OVHeXvLZYLkP2CzN-4)

1. Dropdown container
2. Item
3. Header

> 4. Quick Actions

Reviewer: "a"

## Options

- **Header:** Add a header for the dropdown container. This helps the user to better understand which elements are in the dropdown.
- **Quick Actions:** Add a quick action bar. We typically use quick actions to access common functions such as cut, copy and paste.

> - **Checked:** Mark selected items in the dropdown with a check mark. We typically use check marks when an item can be activated (e.g. for the dropdown "Sort by:" selection: Name, Change date, Create date).

Reviewer: Like that one! 2 small things:
1. Usually "Change date" is actually "Modified date"
2. I would keep this list of options lowercase, but that is maybe personal preference

> - **Submenu:** Add a submenu for a multi-level dropdown. We typically use a submenu when there is a long item list and a varied categorization within the items.

Reviewer: varied categorization = abwechslungsreiche Kategorisierung?

- **Separator:** Add a separator to visually divide items from each other. We normally use a separator to isolate individual elements from a cohesive item list.
- **Icon:** Icons can also be displayed with item labels. We typically use an icon in addition to the label when it has a recognizable benefit.
- **Label:** Set a label for the dropdown item. We typically use short labels including verbs.

> - **Anchor:** An anchor defines where the dropdown is opened. When no anchor is defined the trigger element is used as the anchor.

Reviewer: since you refer to the trigger element: Maybe swap those two, then the reader already knows about the trigger element :)

Reviewer: tbh I personally don't fully understand the difference between the two options 🤔 maybe we can have a chat about this :)

- **Trigger:** The trigger defines which element opens the dropdown. A trigger should also be defined for a dropdown submenu. We typically use a button as the trigger element.

> - **Close Behavior:** Defines which click action closes the dropdown. A dropdown of a submenu is closed with the higher-level dropdown, regardless of its own closing behavior.

Reviewer: Maybe it is sufficient to say "Defines if clicking within and/or outside the dropdown closes the dropdown. A click on a submenu follows the behavior of its parent"
?

(this was just a brief outline to show what I mean - feel free to find a nicer wording for this 😉

Three Options are possible: 
    - Inside: clicking within the dropdown closes the dropdown.
    - Outside: clicking outside the dropdown closes the dropdown.
    - Both: clicking within and outside the dropdown closes the dropdown.

> - **Placement:** Defines where the dropdown appears when it is opened. The position depends on the element that opens the dropdown (e.g. button). By default, the dropdown is displayed at the bottom right of the button. 

Reviewer: is it intentional that you refer to the button at the end of the sentence?

Note that in the beginning of the chapter, you refer to it as a "top element"

> You can choose between different directions (top, bottom, left, right) and two options for the alignment on the button (start, end). When there is not enough space for the selected placement, it will be corrected automatically. The placement of the submenu is always generated automatically. We typically use the default placement option to ensure consistency. 

Reviewer: Maybe this automatic placement is better described in the behavior in context section, and here just briefly described as a reference?
(e.g. "Place a dropdown on the top, bottom, left or right, and the start or end of the top element, though it might change automatically to be in view (described in detail _behavior in context_)

Reviewer: plus, can it be that it is not in one line with the bullet point?

> - **Date Selection:** Use the element [Date dropdown](date-dropdown.md) to get a date selection in the dropdown.

Reviewer: "Date selection" and "date dropdown" :)

Reviewer: Maybe "Place a date dropdown component in the dropdown for a date selection"?
What about time dropdown?

Author: I only used the date dropdown here because it already contains the dropdown. Date piker and date time piker would not automatically be in a dropdown.

I need to ask again if you can simply put a date piker or time piker in a dropdown.

![Dropdown Examples](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2372-2696&mode=design&t=OVHeXvLZYLkP2CzN-4)

## Behavior in context

> - **Text truncation:**  Item and header labels are not truncated. The whole text in the items and headers consists of one line only. The text is cut off after 58 characters with "...".

Reviewer: maybe it's the time - "..." is called a truncation, isn't it?

Author: Here I wanted to clarify where exactly it cuts off and how exactly. Since there is also the variant with (...).

- **Scrollbar:** A dropdown is provided with a scrollbar when the dropdown takes up 50% of the screen.

## States

> Dropdown container items have five states: Default, hover, active, disabled and focused. When a dropdown submenu is active, the dropdown submenu items display an additional dropdown with selectable options.

Reviewer: ... in active state (?)

I would still refer to it as "dropdown items" rather than "dropdown container items"

![Item States](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2343-42235&mode=design&t=OVHeXvLZYLkP2CzN-4)

> 

Reviewer: what do you mean "in advance"?
"deactivate" -> "disable"?

Maybe a do/don't could be trigger elements that are good/bad? (just an idea)

Other than that they are good 👍

Author: in the sense of "do not pack all elements into a dropdown list", but rather structure the elements beforehand and then place them usefully - perhaps you can find a better way of saying this?

I don't really know what a bad trigger element would be, do you have a specific idea?

    - Both: clicking within and outside the dropdown closes the dropdown.
- **Placement:** Defines where the dropdown appears when it is opened. The position depends on the element that opens the dropdown (e.g. button). By default, the dropdown is displayed at the bottom right of the button. 
You can choose between different directions (top, bottom, left, right) and two options for the alignment on the button (start, end). When there is not enough space for the selected placement, it will be corrected automatically. The placement of the submenu is always generated automatically. We typically use the default placement option to ensure consistency. 
- **Date Selection:** Use the component [date dropdown](date-dropdown.md) to get a date selection in the dropdown.

![Dropdown Examples](https://www.figma.com/file/wEptRgAezDU1z80Cn3eZ0o/iX-Pattern-Illustrations?type=design&node-id=2372-2696&mode=design&t=OVHeXvLZYLkP2CzN-4)

## Behavior in context

> - **Text truncation:**  Item and header labels are not truncated. The whole text in the items and headers consists of one line only. The text is cut off after 58 characters.

Reviewer: Stay consistent in wording: label instead of text, and I would use "truncated" also for the last sentence, since it is actually was happening
